src/tree/v2/CachedBranchDataProvider.ts

```diff
@@ -0,0 +1,80 @@
+*  Licensed under the MIT License. See License.txt in the project root for license information.
+*--------------------------------------------------------------------------------------------*/
+
+import { CancellationToken, Event, ProviderResult, TreeItem } from "vscode";
+import { BranchDataProvider, ResourceBase, ResourceModelBase } from "../../api/v2/v2AzureResourcesApi";
+
+type BranchDataProviderBase = BranchDataProvider<ResourceBase, ResourceModelBase>;
+
+type CachedResourceModelBase = ResourceModelBase & {
+    parent?: CachedResourceModelBase;
+}
+
+/**
+ * Wraps a `BranchDataProvider` adding the following:
+ *
+ * - Implements `getParent` by caching calls to `getChildren`.
+ * - in `getTreeItem`, `TreeItem.id` is set to a path based id
+ */
+export class CachedBranchDataProvider<TBranchDataProvider extends BranchDataProviderBase> implements BranchDataProviderBase {
+    constructor(private readonly branchDataProvider: TBranchDataProvider) { }
+
+    async getChildren(element: ResourceModelBase): Promise<ResourceModelBase[] | null | undefined> {
+        const children = await this.branchDataProvider.getChildren(element);
+
+        if (children) {
+            return this.updateItemChildren(element, children);
+        }
+
+        return children;
+    }
+
+    getParent(element: CachedResourceModelBase): ProviderResult<ResourceModelBase> {
+        return element.parent;
+    }
+
+    getResourceItem(element: ResourceBase): ResourceModelBase | Thenable<ResourceModelBase> {
+        return this.branchDataProvider.getResourceItem(element);
+    }
+
+    onDidChangeTreeData: Event<void | ResourceModelBase | ResourceModelBase[] | null | undefined> | undefined = this.branchDataProvider.onDidChangeTreeData;
+
+    async getTreeItem(element: ResourceModelBase): Promise<TreeItem> {
+        const treeItem = await this.branchDataProvider.getTreeItem(element);
+        treeItem.id = this.getId(element);
+        return treeItem;
+    }
+
+    resolveTreeItem(item: TreeItem, element: ResourceModelBase, token: CancellationToken): ProviderResult<TreeItem> {
+        return this.branchDataProvider.resolveTreeItem?.(item, element, token);
+    }
+
+    getId(element: CachedResourceModelBase): string {
+        return this.getPathForItem(element).join('/');
+    }
+
+    private getPathForItem(item: CachedResourceModelBase): string[] {
+        const path: string[] = [];
+
+        let currentItem: CachedResourceModelBase | undefined = item;
+
+        while (currentItem) {
+            const nextItem: CachedResourceModelBase | undefined = currentItem.parent;
+
+            if (currentItem.id) {
+                path.push(currentItem.id);
+            }
+
+            currentItem = nextItem;
+        }
+
+        return path.reverse();
+    }
+
+    private updateItemChildren(parent: ResourceModelBase, children: ResourceModelBase[]): CachedResourceModelBase[] {
+        (children as CachedResourceModelBase[]).forEach(child => child.parent = parent);
```

Comment:
This caching alters the model's returned by the original BDP (i.e. adding a `parent`) property. Do we expect users of this caching layer to use that `parent` property separately from RG's purpose of generating paths, IDs, and parents for find/reveal?  If not (which I feel like will be the case), then there's a chance of collision with that BDP as it's not going to expect models to be modified.  In addition, I don't think we restrict models from appearing twice in a tree.  In that case, it might be better for this caching layer to return its own wrapped model (i.e. with the original model and `parent` properties as siblings).  The quirk about that approach is that such types would need to be unwrapped *twice*.  (We could have `BrandDataProviderItem.unwrap()` repeatedly call `unwrap()` until no more `unwrap()` functions exist on the inner model.

Another approach would be to store the child -> parent mapping in a separate type (but then requires more set/clear management).

Reply:
If we change the property name to something more obscure like `_cachedParent`, and make it clear in the docs that using `CachedBranchDataProvider` modifies the models by adding this property, then I think it's ok. Especially considering the main reason to use `CachedBranchDataProvider` is to implement `getParent`, and it's optional.

Reply:
I think that's probably a reasonable compromise (and can always be changed if/when it ever becomes an issue).  I might pick a name that's more "namespaced", though, something like `_rgApiCachedParent`.

Reply:
Ok, I changed the name to `_rgApiCachedParent`.